src/dcap_provider.cpp

```diff
@@ -707,21 +769,83 @@ static std::string build_enclave_id_url(
     return qe_id_url.str();
 }
 
+static std::unique_ptr<std::vector<uint8_t>> try_cache_get(
+    const std::string& cert_url)
+{
+    try 
+    {
+        return local_cache_get(cert_url);
+    }
+    catch (std::runtime_error& error)
+    {
+        log(SGX_QL_LOG_WARNING, "Unable to access cache: %s", error.what());
+        return nullptr;
+    }
+}
+
+static std::string get_issuer_chain_cache_name(std::string url)
+{
+    return url + "IssuerChain";
+}
+
 static quote3_error_t get_collateral(
+    CollateralTypes collateral_type,
     std::string url,
     const char header_name[],
     std::vector<uint8_t>& response_body,
     std::string& issuer_chain,
     const std::string* const request_body = nullptr)
 {
+    quote3_error_t retval = SGX_QL_ERROR_UNEXPECTED;
+    std::string friendly_name = get_collateral_friendly_name(collateral_type);
     try
     {
-        const auto crl_operation = curl_easy::create(url, request_body);
-        crl_operation->perform();
-        response_body = crl_operation->get_body();
+        std::string issuer_chain_cache_name = get_issuer_chain_cache_name(url);
+        if (auto cache_hit_collateral = try_cache_get(url))
+        {
+            if (auto cache_hit_issuer_chain = try_cache_get(issuer_chain_cache_name))
+            {
+                log(SGX_QL_LOG_INFO,
+                    "Fetching %s from cache: '%s'.",
+                    friendly_name.c_str(),
+                    url.c_str());
+                response_body = *cache_hit_collateral;
+                issuer_chain = std::string(cache_hit_issuer_chain->begin(), cache_hit_issuer_chain->end());
+                return SGX_QL_SUCCESS;
+            }
+        }
+
+        log(SGX_QL_LOG_INFO,
+            "Fetching %s from remote server: '%s'.",
+            friendly_name.c_str(),
+            url.c_str());
+
+        const auto curl_operation = curl_easy::create(url, request_body);
+        curl_operation->perform();
+        response_body = curl_operation->get_body();
         auto get_header_operation =
-            get_unescape_header(*crl_operation, header_name, &issuer_chain);
-        return convert_to_intel_error(get_header_operation);
+            get_unescape_header(*curl_operation, header_name, &issuer_chain);
+
+        retval = convert_to_intel_error(get_header_operation);
+
+        // Update the cache if needed
+        time_t expiry = 0;

```

Comment:
You cache even if there's an error? But the cache doesn't seem to include the error, so I'm not 100% sure why you'd cache on failure here.

Reply:
Thanks! Definitely don't want to be caching on error - fixed